### PR TITLE
Add self-supervised 3D-Var-based AI data assimilation 

### PR DESCRIPTION
**Pull request by SOHAMPAL23**
# Pull Request

- Add self-supervised AI-based data assimilation using 3D-Var loss

## Description

- This PR introduces a self-supervised AI-based data assimilation prototype

- The implementation replaces supervised learning with a physics-based 3D-Var cost function, allowing the model to learn directly from:

- Sparse/noisy observations

## How Has This Been Tested?

_Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce.
Please also list any relevant details for your test configuration_

- [ ] Yes

_If your changes affect data processing, have you plotted any changes? i.e. have you done a quick sanity check?_

- [ ] Yes

## Checklist:

- [ ] My code follows [OCF's coding style guidelines](https://github.com/openclimatefix/.github/blob/main/coding_style.md)
- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked my code and corrected any misspellings
https://github.com/openclimatefix/graph_weather/issues/173

### CODE REVIEW

File: tests/test_ai_assimilation.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+sys.path.insert(0, ai_assimilation_path)
+
+# Import the modules directly (avoiding package import issues)
+import model
+import loss
+import data
+import training
+
+
+class TestAIAssimilation(unittest.TestCase):
```
Comment (jacobbieker) [inline tests/test_ai_assimilation.py]: We use pytest, not unit test, so please change to pytest.

File: tests/test_ai_assimilation.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
+import sys
+import os
+import torch
+import unittest
+
+# Add the AI assimilation module directory to the path
+ai_assimilation_path = os.path.join(
```
Comment (jacobbieker) [inline tests/test_ai_assimilation.py]: Don't do this, just import correctly

File: tests/test_ai_assimilation.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+        # Calculate loss
+        total_loss = loss_fn(analysis, first_guess, observations)
+
+        # Verify loss properties
+        self.assertEqual(total_loss.dim(), 0, "Loss should be a scalar tensor")
+        self.assertGreaterEqual(total_loss.item(), 0, "Loss should be non-negative")
+        self.assertFalse(torch.isnan(total_loss).any().item(), "Loss should not contain NaN values")
+        self.assertFalse(torch.isinf(total_loss).any().item(), "Loss should not contain Inf values")
+
+    def test_synthetic_data_generation(self):
```
Comment (jacobbieker) [inline tests/test_ai_assimilation.py]: We shouldn't be generating synthetic data in this repo, and if using synthetic data for tests, shouldn't need to test it

File: tests/test_ai_assimilation.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+    if result.wasSuccessful():
+        print("ALL TESTS PASSED! AI assimilation module is working correctly.")
+    else:
+        print(" SOME TESTS FAILED! Please check the AI assimilation module.")
+        print(f"Failures: {len(result.failures)}, Errors: {len(result.errors)}")
+
+    return result.wasSuccessful()
+
+
+if __name__ == "__main__":
```
Comment (jacobbieker) [inline tests/test_ai_assimilation.py]: Like I said before, don't use main

File: tests/test_ai_assimilation.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+        self.assertIsInstance(train_loss, float, "Training loss should be a float")
+        self.assertFalse(
+            torch.isnan(torch.tensor(train_loss)).any().item(), "Training loss should not be NaN"
+        )
+        self.assertFalse(
+            torch.isinf(torch.tensor(train_loss)).any().item(), "Training loss should not be Inf"
+        )
+
+
+def run_tests():
```
Comment (jacobbieker) [inline tests/test_ai_assimilation.py]: This should run with pytest, so don't need this.

File: graph_weather/models/ai_assimilation/model.py [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+        # Concatenate first-guess and observations
+        combined_input = torch.cat([first_guess, observations], dim=-1)
+
+        # Pass through the network to get the analysis
+        analysis = self.network(combined_input)
+
+        return analysis
+
+
+class SimpleAIBasedAssimilationNet(nn.Module):
```
Comment (jacobbieker) [inline graph_weather/models/ai_assimilation/model.py]: Like I said in the other one, remove this please.

File: graph_weather/models/ai_assimilation/training.py [hunk 7]
```diff
@@ -0,0 +1,415 @@
...
+from typing import Any, Dict, Optional, Tuple
+
+import matplotlib.pyplot as plt
+import torch
+from torch.nn import Module
+from torch.optim import Adam
+from torch.optim.lr_scheduler import ReduceLROnPlateau
+from tqdm import tqdm
+
+current_dir = os.path.dirname(os.path.abspath(__file__))
```
Comment (jacobbieker) [inline graph_weather/models/ai_assimilation/training.py]: Don't do this kind of thing, just import things correctly.

File: graph_weather/models/ai_assimilation/training.py [hunk 8]
```diff
@@ -0,0 +1,415 @@
...
+loss_file = os.path.join(current_dir, "loss.py")
+loss_spec = importlib.util.spec_from_file_location("loss", loss_file)
```
Comment (jacobbieker) [inline graph_weather/models/ai_assimilation/training.py]: Just import the files, this isn't a good way of doing it

**Comment by Sidharth1743** [comment]
@SOHAMPAL23 use the review button. don't tag him. it will be notified in that way itself. thanks:)

**Comment by SOHAMPAL23** [comment]
@Sidharth1743  Thanks for letting me know that. 

So Till  he reviews it, I can work on other issues

**Comment by SOHAMPAL23** [comment]
@jacobbieker  Can you give review on this PR
